Print log for aborted connections

nemisj opened the pull request:
It's possible that requests takes so long that it's either aborted by the browser or by the http.Server.
In both cases, there is no log call saying that the connection is dead, which gives the feeling in the logs, that request is still in progress.

davidmarkclements commented:
Yes! @nemisj v. nice

1. Would you mind running comparison benchmark?
2. There's a couple of linting issues

davidmarkclements commented on logger.js:
```diff
@@ -27,6 +27,7 @@ function pinoLogger (opts, stream) {
   function onResFinished (err) {
     this.removeListener('finish', onResFinished)
     this.removeListener('error', onResFinished)
+    this.req.removeListener('aborted', onReqAborted);
```
Can we get away with not removing aborted listener?

If req is done, aborted won't fire and the whole object is heading for gc. I ask because removeListener isn't cheap

marcbachmann replied:
Good idea. Why don't we do the same for the response?

marcbachmann commented on logger.js:
```diff
@@ -46,13 +47,20 @@ function pinoLogger (opts, stream) {
   }
 
+  function onReqAborted() {
+    this.res.statusCode = 0;
```
@davidmarkclements how do you feel about overwriting the `statusCode` in here?
I wouldn't do it.

nemisj replied:
@marcbachmann, if we don't do it there, then the log will be with 200 statusCode, that's why it's here

nemisj commented:
@davidmarkclements 

1. how can i do that?
2. don't see any lint config, what rules do you use? what should I change?

davidmarkclements commented:
@nemisj 

1. linting uses standard (http://npm.im/standard) - run npm test to check
2. yah sorry benchmarking isn't obvious on this module, run `node benchmarks/pino-server.js` and then use [autocannon](http://npm.im/autocannon) to bench the http server, - bench it before the change and after (pref three times each, then take avg) 
3. could you also add tests (see tests.js)

nemisj commented:
@davidmarkclements, @marcbachmann 
Guys, I've linted, added tests and removed `removeListener`
I haven't tested it yet in a real wordl use-case after deleting 'removeListener', so i will do it to see if it's not leaking and also I will do benchmarking.

mcollina commented on logger.js:
```diff
@@ -46,13 +46,22 @@ function pinoLogger (opts, stream) {
+    // it's possible that browser aborts connection, or http-server because of timeout
+    req.on('aborted', function () {
+      onReqAborted(res)
+    })
```
we should avoid creating this closure if possible, not sure if there is a way. Maybe we can attach the response to the request (if it is not already there)

nemisj commented:
@davidmarkclements how much connections do you use as default for benchmarking? any other options?

nemisj commented:
statusCode is not working, indeed. http_server starts to puking if there are some unflushed data.

```
RangeError: Invalid status code: 0
    at ServerResponse.writeHead (_http_server.js:192:11)
```

But 200 is also not nice in the log, any ideas?

mcollina commented:
IMHO we shouldn't touch it. However, we should detect if the request was aborted, and correctly _not_ printing the status code in the serializer.
@davidmarkclements do you think it could solve your issues?

nemisj commented:
@mcollina, why do you think we shouldn't touch it? we touch already request object and response object with augmenting it and statusCode in this case is something what can be used as a assertion also outside the logger. I see there is a 408 http code - http://www.checkupdown.com/status/E408.html i guess this is the correct one which should be used instead. What do you think?

mcollina commented:
@nemisj we shouldn't touch t because the user might change things. But you are right, setting 408 is correct.

nemisj commented:
@mcollina i will put 408 in the code

marcbachmann commented on logger.js:
```diff
@@ -46,13 +46,22 @@ function pinoLogger (opts, stream) {
     req.id = genReqId(req)
     req.log = res.log = logger.child({req: req})
     res.startTime = Date.now()
+    if (!req.res) { req.res = res }
```
I don't get why we have to check for `req.res` before setting it.
IMO we can remove that. Just use `req.res = res` directly.

marcbachmann replied:
i just checked express. They are setting req.res, the node http server isn't.

mcollina replied:
Let's leave as it is then.
